### PR TITLE
Add Animal Options To Valid Animal Viewsheet

### PR DESCRIPTION
**Pull request by LightHardt**
Continuation of https://github.com/DFHack/scripts/pull/961

This will resolve https://github.com/DFHack/dfhack/issues/4043

Integration of script into zone.lua

### CODE REVIEW

File: plugins/lua/zone.lua [hunk 1]
```diff
@@ -1216,10 +1216,188 @@ function RetireLocationOverlay:confirm_retire()
...
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    return not dfhack.units.isPet(unit)
+end
+
+local function is_avail_adoption()
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    -- Make sure not already a pet and not a cat as they adopt owner
+    return not dfhack.units.isPet(unit) and unit.race ~= 171
```
Comment (myk002) [inline plugins/lua/zone.lua:1252]: This won't be true for all worlds. Race IDs will change according to the mods you have loaded.

Is there a race trait we can check against? If not, you might have to compare against the string id (`CAT`)

File: plugins/lua/zone.lua [hunk 2]
```diff
@@ -1216,10 +1216,188 @@ function RetireLocationOverlay:confirm_retire()
...
+}
+
+-- The above function already handles checking if valid unit
+-- so just set slaughter flag
+function AnimalActionsWidget:set_slaughter_flag(option)
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    if not unit then return end
+
+    if option then
```
Comment (myk002) [inline plugins/lua/zone.lua:1271]: Can just have `unit.flags2.slaughter = option`

Comment (myk002) [reply plugins/lua/zone.lua:1271]: Similar for other callbacks

File: plugins/lua/zone.lua [hunk 3]
```diff
@@ -1216,10 +1216,188 @@ function RetireLocationOverlay:confirm_retire()
...
+        unit.flags3.available_for_adoption = true
+    else
+        unit.flags3.available_for_adoption = false
+    end
+end
+
+-- Check current flag status of animal to dynamically set on/off
+function AnimalActionsWidget:get_butcher(unit)
```
Comment (myk002) [inline plugins/lua/zone.lua:1305]: These get functions are unused and can be removed

File: plugins/lua/zone.lua [hunk 4]
```diff
@@ -1216,10 +1216,188 @@ function RetireLocationOverlay:confirm_retire()
         curry(retire, location))
 end
 
+-- -------------------
+-- AnimalActionsWidget
+--
+
+local YES = 'Yes'
```
Comment (myk002) [inline plugins/lua/zone.lua:1223]: These constants are of questionable utility

**Comment by myk002** [comment]
Looking good! Don't forget the changelog entry!

File: plugins/lua/zone.lua (hunk 1 above)
Comment (LightHardt) [reply plugins/lua/zone.lua]: I found a method that caravan uses and slightly modified it and seems to work.

File: plugins/lua/zone.lua (hunk 2 above)
Comment (LightHardt) [reply plugins/lua/zone.lua:1268]: made change

File: plugins/lua/zone.lua (hunk 3 above)
Comment (LightHardt) [reply plugins/lua/zone.lua]: Removed functions

File: plugins/lua/zone.lua (hunk 4 above)
Comment (LightHardt) [reply plugins/lua/zone.lua]: Removed constants

**Comment by LightHardt** [comment]
Thank you for all the help this process would have been much slower without it!

File: plugins/lua/zone.lua [hunk 5]
```diff
@@ -1216,10 +1216,154 @@ function RetireLocationOverlay:confirm_retire()
...
+    return unit
+        and dfhack.units.isFortControlled(unit)
+        and dfhack.units.isAlive(unit)
+        and dfhack.units.isAnimal(unit)
+end
+
+local function is_geldable()
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    return dfhack.units.isGeldable(unit)
```
Comment (myk002) [inline plugins/lua/zone.lua:1235]:
```suggestion
    return unit and dfhack.units.isGeldable(unit)
```

File: plugins/lua/zone.lua [hunk 6]
```diff
@@ -1216,10 +1216,154 @@ function RetireLocationOverlay:confirm_retire()
...
+end
+
+local function is_not_pet()
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    return not dfhack.units.isPet(unit)
```
Comment (myk002) [inline plugins/lua/zone.lua:1241]:
```suggestion
    return not unit or not dfhack.units.isPet(unit)
```

File: plugins/lua/zone.lua [hunk 7]
```diff
@@ -1216,10 +1216,154 @@ function RetireLocationOverlay:confirm_retire()
...
+end
+
+local function is_avail_adoption()
+    local unit = dfhack.gui.getSelectedUnit(true)
```
Comment (myk002) [inline plugins/lua/zone.lua:1245]:
```suggestion
    local unit = dfhack.gui.getSelectedUnit(true)
    if not unit then return false end
```

**Comment by LightHardt** [comment]
Wondering if i should rename the `Tame` label to `Train`. Not sure.

File: docs/changelog.txt [hunk 8]
```diff
@@ -54,6 +54,7 @@ Template for new versions:
 ## New Tools
 
 ## New Features
+- ``zone.animal_actions``: Add overlay to tamed animals of civ for butchering/gelding/adoption/tame(assign trainer) options in their "Overview" tab
```
Comment (myk002) [inline docs/changelog.txt:57]:
```suggestion
- `zone`: Add overlay for toggling butchering/gelding/adoption/taming options in animal "Overview" tabs
```

File: plugins/lua/zone.lua [hunk 9]
```diff
@@ -1216,10 +1216,186 @@ function RetireLocationOverlay:confirm_retire()
...
+
+    local creature_id = df.creature_raw.find(unit.race).creature_id
+    -- Make sure not already a pet and not a cat as they adopt owner
+    return not dfhack.units.isPet(unit) and creature_id ~= 'CAT'
```
Comment (myk002) [inline plugins/lua/zone.lua:1249]:
```suggestion
    if dfhack.units.isPet(unit) return false end
    
    local raw = df.creature_raw.find(unit.race)
    if not raw then return false end
    
    -- cats adopt owners; owners can't adopt cats
    return raw.creature_id ~= 'CAT'
```

File: plugins/lua/zone.lua [hunk 10]
```diff
@@ -1216,10 +1216,186 @@ function RetireLocationOverlay:confirm_retire()
...
+function AnimalActionsWidget:set_slaughter_flag(option)
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    if not unit then return end
+
+    if option then
+        unit.flags2.slaughter = option
+    else
+        unit.flags2.slaughter = option
+    end
```
Comment (myk002) [inline plugins/lua/zone.lua:1278]:
```suggestion
    unit.flags2.slaughter = option
```

File: plugins/lua/zone.lua [hunk 11]
```diff
@@ -1216,10 +1216,186 @@ function RetireLocationOverlay:confirm_retire()
...
+function AnimalActionsWidget:set_geld_flag(option)
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    if not unit then return end
+
+    if option then
+        unit.flags3.marked_for_gelding = option
+    else
+        unit.flags3.marked_for_gelding = option
+    end
```
Comment (myk002) [inline plugins/lua/zone.lua:1291]:
```suggestion
    unit.flags3.marked_for_gelding = option
```

File: plugins/lua/zone.lua [hunk 12]
```diff
@@ -1216,10 +1216,186 @@ function RetireLocationOverlay:confirm_retire()
...
+function AnimalActionsWidget:set_adoption_flag(option)
+    local unit = dfhack.gui.getSelectedUnit(true)
+
+    if not unit then return end
+
+    if option then
+        unit.flags3.available_for_adoption = option
+    else
+        unit.flags3.available_for_adoption = option
+    end
```
Comment (myk002) [inline plugins/lua/zone.lua:1304]:
```suggestion
    unit.flags3.available_for_adoption = option
```

**Review by myk002 — COMMENTED** [review]
love it. I suggested a few cleanups. The guarding against nil units is good practice, even though the current setup *should* prevent them from ever being nil. If things change in the future, though, it's best not to have an error.

**Comment by myk002** [comment]
> Wondering if i should rename the `Tame` label to `Train`? Not sure.

I'd go with "Tame" since we're not specifically training for war/hunting (though that's a potential future improvement to offer those options)

> Also `isMarkedForTraing` is marking my war Giant Leopard as `Yes` should i also add a `isDomesticated` check for that? Leaning towards no since it is indeed marked for training and it being domesticated already disables the button so really it would be a visual matter of wanting it to say either No or Yes but curious what you think.

I'd go with keeping it as it is -- it does have a trainer assigned, after all.

File: plugins/lua/zone.lua (hunk 5 above)
Comment (LightHardt) [reply plugins/lua/zone.lua]: Understood

File: plugins/lua/zone.lua (hunk 10 above)
Comment (LightHardt) [reply plugins/lua/zone.lua:1283]: I now see what you were trying to say my bad

**Comment by myk002** [comment]
> I'd go with "Tame" since we're not specifically training for war/hunting (though that's a potential future improvement to offer those options)

Looking at this in-game..
image

the pig is tame (domesticated) but we say `Tame: No`

Maybe change that to "has trainer"?

**Comment by myk002** [comment]
also looking in-game, maybe line up the labels so it's more like:
```lua
Butcher: No
Geld:    No
Adopt:  Yes
Tame     No
```

**Comment by myk002** [comment]
Finally, it seems that there are some invariants that we have to enforce:

https://github.com/DFHack/dfhack/assets/977482/2f92376b-3302-4c8e-aadb-e7f733b50001

Marking for slaughtering unmarks adoption and gelding
marking for adoption unmarks slaughter
marking for gelding unmarks slaughter

**Comment by LightHardt** [comment]
> Marking for slaughtering unmarks adoption and gelding marking for adoption unmarks slaughter marking for gelding unmarks slaughter

Thats new behavior I tested earlier and that wasn't an issue I wonder which commit it was, ill have to look.

**Comment by myk002** [comment]
> Thats new behavior I tested earlier and that wasn't an issue I wonder which commit it was, ill have to look.

What I was screen capturing was vanilla behavior -- I'm saying we should make sure to emulate it

**Comment by LightHardt** [comment]
> > Thats new behavior I tested earlier and that wasn't an issue I wonder which commit it was, ill have to look.
> 
> What I was screen capturing was vanilla behavior -- I'm saying we should make sure to emulate it

Lol i got it backwards my bad, i see what you're saying i was thinking a shared variable but windering how that would work with multiple units.

**Comment by myk002** [comment]
I think it can just be enforced in the "set" functions -- when we're setting  one of them to true, unconditionally set the ones that should be unmarked to false

**Comment by myk002** [comment]
e.g.: 

```lua
function AnimalActionsWidget:set_slaughter_flag(option)
    local unit = dfhack.gui.getSelectedUnit(true)
    if not unit then return end
    if option then
        self:set_geld_flag(false)
        self:set_adopt_flag(false)
    end
    unit.flags2.slaughter = option
end
```

**Comment by LightHardt** [comment]
Switched to `Has Trainer`
Padded labels so all even
Emulate vanilla method with being only able to set 1 at a time (thank you)

Let me know what you think!

File: plugins/lua/zone.lua [hunk 13]
```diff
@@ -1285,6 +1292,11 @@ function AnimalActionsWidget:set_geld_flag(option)
 
     if not unit then return end
 
+    if option then
+        self:set_slaughter_flag(false)
+        self:set_adoption_flag(false)
```
Comment (myk002) [inline plugins/lua/zone.lua:1297]: Setting geld does not unset adoption (which makes sense)

File: plugins/lua/zone.lua [hunk 14]
```diff
@@ -1294,6 +1306,11 @@ function AnimalActionsWidget:set_adoption_flag(option)
 
     if not unit then return end
 
+    if option then
+        self:set_geld_flag(false)
```
Comment (myk002) [inline plugins/lua/zone.lua:1310]: likewise, setting adopt does not affect the geld flag

**Review by myk002 — APPROVED** [review]

**Comment by myk002** [comment]
This looks so nice. Thank you very much for contributing it!

**Comment by LightHardt** [comment]
> This looks so nice. Thank you very much for contributing it!

Thanks for helping me pretty much all day with it, it's much appreciated!

**Pull request merged**